decode.go

```diff
@@ -446,22 +440,49 @@ func parseIntLiteral(value string) (intLit, error) {
 	return lit, err
 }
 
-// Searches for the field with the given 'key' first by Name, then by JSONName,
+func (u *unmarshaler) getExtResolver() protoregistry.ExtensionTypeResolver {
+	if u.options.Resolver != nil {
+		return u.options.Resolver
+	}
+	return protoregistry.GlobalTypes
+}
+
+func (u *unmarshaler) getResolver() protoregistry.MessageTypeResolver {
+	if u.options.Resolver != nil {
+		return u.options.Resolver
+	}
+	return protoregistry.GlobalTypes
+}
+
+// findField searches for the field with the given 'key' by extension type, JSONName, TextName,
 // and finally by Number.
-func findField(key string, fields protoreflect.FieldDescriptors) protoreflect.FieldDescriptor {
-	if field := fields.ByName(protoreflect.Name(key)); field != nil {
-		return field
+func (u *unmarshaler) findField(key string, msgDesc protoreflect.MessageDescriptor) (protoreflect.FieldDescriptor, error) {
+	fields := msgDesc.Fields()
+	if strings.HasPrefix(key, "[") && strings.HasSuffix(key, "]") {
+		extName := protoreflect.FullName(key[1 : len(key)-1])
+		extType, err := u.getExtResolver().FindExtensionByName(extName)
+		if err != nil {
+			return nil, err
+		}
+		result := extType.TypeDescriptor()
+		if !msgDesc.ExtensionRanges().Has(result.Number()) || result.ContainingMessage().FullName() != msgDesc.FullName() {
+			return nil, fmt.Errorf("message %v cannot be extended by %v", msgDesc.FullName(), result.FullName())
+		}
```

Comment:
nit: the first part of the check (that the extension tag is in a valid range) should not be necessary. The protobuf-go runtime won't create an extension descriptor that is invalid in this way.

Reply:
This is identical to json, and will give you a NodeError instead of a error with out a Node for context